kahuna/public/js/image/service.js

```diff
@@ -44,3 +45,274 @@ imageService.factory('imageService', ['editsService', function(editsService) {
 
     return image => forImage(image);
 }]);
+
+
+imageService.factory('unique', function() {
+    return arr => {
+        return arr.reduce((prev, curr) =>
+            prev.indexOf(curr) !== -1 ? prev : prev.concat([curr]), []);
+    };
+});
+
+// takes an array of objects and turns it into an object with an array of unique values
+// e.g. [{ a: 1, b: 2 }, { a: 2, b: 2, c: 3 }] => { a: [1,2], b: [2], c: [3] }
+imageService.factory('reduceObjectsToArrays', [function() {
+
+    return function (objects, presetKeys) {
+        return objects.reduce((prev, curr) => {
+            const keys = presetKeys || Object.keys(prev).concat(Object.keys(curr));
+            keys.forEach(field => {
+                prev[field] = (prev[field] || []).concat([curr[field]]);
+            });
+            return prev;
+        }, {});
+    }
+
+}]);
+
+imageService.factory('uniqueArrayKeyValues', ['unique', function(unique) {
+    return function (object) {
+        return Object.keys(object).reduce((prev, field) => {
+            prev[field] = unique(object[field]).filter(v => v);
+            return prev;
+        }, {});
+    }
+}])
+
+imageService.factory('labelsService', ['$q', 'unique', function($q, unique) {
+    const updates$ = new Rx.Subject();
+
+    function getLabels(image) {
+        return image.data.userMetadata.data.labels;
+    }
+
+    function labelsService(images$) {
+        const labels$ = images$.map(uniqueLabels);
+
+        return { add, remove, updates$, labels$ };
+
+        function uniqueLabels(images) {
+            return unique(images.reduce((prev, curr) =>
+                prev.concat(curr.data.userMetadata.data.labels.data.map(l => l.data)), []));
+        }
+
+        function remove(label) {
+            if (label) {
+                const edits = images$.getValue().map(image => {
+                    const labelResource =
+                        getLabels(image).data.find(labelResource => labelResource.data === label);
+
+                    if (labelResource) {
+                        return labelResource.delete().then(labels => {
+                            image.data.userMetadata.data.labels = labels;
+                            return image;
+                        });
+                    }
+                }).filter(v => v);
+
+                $q.all(edits).then(images => updates$.onNext(images));
+            }
+        }
+
+        function add(label) {
+            if (label) {
+                const edits = images$.getValue().map(image => getLabels(image).
+                    post({ data: [label] }).
+                    then(labels => {
+                        image.data.userMetadata.data.labels = labels;
+                        return image;
+                    })
+                );
+
+                $q.all(edits).then(images => updates$.onNext(images));
+            }
+        }
+    }
+
+    return labelsService;
+
+}]);
+
+
+imageService.factory('metadataService',
```

Comment:
It feels a bit strange that everything is now on the `imageService` rather than being standalone services.